### PR TITLE
checksum extensions light

### PR DESCRIPTION
**Pull request by alikon**
light porting of #17555 from 4.0 to 3.8

### Summary of Changes
 trigger the checksum check on install/update
 if the update server manifest have an hash tag like these:

hashmanifest

only return:
 
- notice when there is no hash tag in the update manifest
-  warning when a checksum ha failed
-  message when checksum is OK

the installation/update works as before 

### Testing Instructions
see #17555
the main difference is it install anyway so there is no force button

### Expected result
checking the integrity of the package with: SHA256,SHA384,SHA512 hash algos

### Actual result
no check is made

### Documentation Changes Required
 new tags `<sha256><sha256/><sha384><sha384/><sha512><sha512/>` on update server manifest

### CODE REVIEW

**Comment by mbabker** [comment]
1) This should be incorporated into com_joomlaupdate as well so core can use it.

2) We need to make a decision on what hashes to support.  In 2017, md5 and sha1 aren't great options.

File: libraries/src/Installer/InstallerHelper.php [hunk 1]
```diff
@@ -333,4 +333,46 @@ public static function splitSql($query)
...
+			if ($update->get($hash, false))
+			{
+				$hash_package = hash_file($hash, $packagefile);
+				$hash_remote  = $update->$hash->_data;
+
+				$hashOnFile = true;
+
+				if ($hash_package !== $hash_remote)
+				{
+					return false;
```
Comment (mbabker) [inline libraries/src/Installer/InstallerHelper.php]: Can we use an integer based return instead of true/false/null?

**Comment by brianteeman** [comment]
> We need to make a decision on what hashes to support. In 2017, md5 and sha1 aren't great options.

sha256 ?

**Comment by mbabker** [comment]
I can't say I know much about sha256's security off hand.

File: libraries/src/Installer/InstallerHelper.php (hunk 1 above)
Comment (alikon) [reply libraries/src/Installer/InstallerHelper.php]: sure

**Comment by alikon** [comment]
> This should be incorporated into com_joomlaupdate as well so core can use it.

sure, can we do in a separate pr ?

> We need to make a decision on what hashes to support. In 2017, md5 and sha1 aren't great options.

sha256 should be OK nowdays ...or not  :confused:

File: libraries/src/Installer/InstallerHelper.php [hunk 2]
```diff
@@ -333,4 +333,46 @@ public static function splitSql($query)
...
+			if ($update->get($hash, false))
+			{
+				$hash_package = hash_file($hash, $packagefile);
+				$hash_remote  = $update->$hash->_data;
+
+				$hashOnFile   = true;
+
+				if ($hash_package !== $hash_remote)
+				{
+					return 0;
```
Comment (mbabker) [inline libraries/src/Installer/InstallerHelper.php]: OK, one more request, so reading the code is a bit simpler.  Add class constants please?

```php
const HASH_NOT_VALIDATED = 0;
const HASH_VALIDATED = 1;
const HASH_NOT_PROVIDED = 2;
```

Comment (alikon) [reply libraries/src/Installer/InstallerHelper.php]: added CONST as requested

**Review by mbabker — APPROVED** [review]

**Comment by mbabker** [comment]
I already know what he's going to nudge toward but I'm going to ask an opinion anyway.

@paragonie-scott are there any major issues to be aware of with using sha256 or sha512 with `hash_file()` as a means for validating a checksum of an extension (or our core) package as a short-to-mid term solution?  (Clearly we could/should continue to work for better options, so right now this is basically a starting point; the one thing this wouldn't address is a compromise of anyone's distribution platform)

**Comment by alikon** [comment]
in the gsoc 17 EEM  meeting yesterday, i've discovered that there are some 3dp extensions that use
md5/sha1 like ARS, so apart the discussion about weakness,  should we consider to support  those weak algos?
one step forward (sha512) and one step back (md5,sha1)

**Comment by mbabker** [comment]
I'm well aware that ARS generates them.  That still doesn't necessarily make md5/sha1 good to introduce into new APIs in the core libraries in 2017 as a means for validating package integrity.

**Comment by paragonie-scott** [comment]
> are there any major issues to be aware of with using sha256 or sha512 with hash_file() as a means for validating a checksum of an extension (or our core) package as a short-to-mid term solution? 

There's a length extension attack, but that doesn't break preimage security, and since it's an unkeyed hash, it's not giving the attacker a capability they don't already have.

However, if you want to be conservative here, follow Composer's example and go with SHA384 specifically.

**Comment by alikon** [comment]
so we can go with: sha256,sha384,sha512  ?

**Comment by mbabker** [comment]
That seems fair to me.

**Comment by alikon** [comment]
now hash supported are:

- sha256
- sha348
- sha512

File: administrator/language/en-GB/en-GB.com_installer.ini [hunk 3]
```diff
@@ -64,6 +64,9 @@ COM_INSTALLER_HEADING_UPDATESITE_NAME_ASC="Update Site ascending"
 COM_INSTALLER_HEADING_UPDATESITE_NAME_DESC="Update Site descending"
 COM_INSTALLER_HEADING_UPDATESITEID="ID"
 COM_INSTALLER_INSTALL_BUTTON="Install"
+COM_INSTALLER_INSTALL_CHECKSUM_NOT_FOUND="There were no checksums provided in the package."
+COM_INSTALLER_INSTALL_CHECKSUM_WRONG="File Checksum Failed"
+COM_INSTALLER_INSTALL_CHECKSUM_CORRECT="File Checksum OK"
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.com_installer.ini]: Sort in alpha order.

File: libraries/src/Installer/InstallerHelper.php [hunk 4]
```diff
@@ -333,4 +357,46 @@ public static function splitSql($query)
 
 		return $db->splitSql($query);
 	}
+
+	/**
+	 * Return the result of the checksum of a package and the SHA1/MD5 tags in the update server manifest
```
Comment (Quy) [inline libraries/src/Installer/InstallerHelper.php]: Update `SHA1/MD5`.

Comment (alikon) [reply libraries/src/Installer/InstallerHelper.php]: ops legacy ;)

File: administrator/language/en-GB/en-GB.com_installer.ini (hunk 3 above)
Comment (alikon) [reply administrator/language/en-GB/en-GB.com_installer.ini]: done
thanks

**Comment by anibalsanchez** [comment]
I have tested this item :white_check_mark: successfully on 2f3425f749c3ca6d5aa364181217dd401d7322c6<br /><br />Tested OK, as part of #17632 

For more info about the test and comment: https://issues.joomla.org/tracker/joomla-cms/17632#event-3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17619">issues.joomla.org/tracker/joomla-cms/17619</a>.</sub>

**Comment by NunoLopesPT** [comment]
I have tested this item :white_check_mark: successfully on 2f3425f749c3ca6d5aa364181217dd401d7322c6<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17619">issues.joomla.org/tracker/joomla-cms/17619</a>.</sub>

**Comment by ghost** [comment]
RTC after two successful tests.

**Comment by mbabker** [comment]
@wilsonge Can we get this and https://github.com/joomla/joomla-cms/pull/17632 merged into the 3.9 branch?  IMO now would be a good time for the code to hit core and start documenting the hell out of the new functionality via the docs wiki and a targeted blog post on the dev network.

**Comment by wilsonge** [comment]
LGTM :)

**Comment by wilsonge** [comment]
Well bugger - didn't see this was against staging. Reverted with #17619 and going to manually re-patch against 3.9 now

**Comment by wilsonge** [comment]
Merged second time lucky with https://github.com/joomla/joomla-cms/commit/f18c20a46ee2eae17d15df3ed5e7c069c94c4820

**Comment by alikon** [comment]
@wilsonge thanks
can you consider to evaluate/merge to 3.9 #17632 too after conflict has been resolved ?

**Pull request merged**